File: server/core/src/main/scala/org/oliverlittle/clusterprocess/model/table/TableTransformation.scala
package org.oliverlittle.clusterprocess.model.table

import org.oliverlittle.clusterprocess.model.field.expressions.{NamedFieldExpression, FieldExpression, AggregateExpression}
import org.oliverlittle.clusterprocess.model.field.comparisons.FieldComparison
import org.oliverlittle.clusterprocess.model.table.field._
import org.oliverlittle.clusterprocess.table_model

import scala.util.Try
import scala.util.hashing.MurmurHash3

sealed trait Assembler:
	/**
	  * Combines a set of partial results into a final output
	  *
	  * @param data An iterable containing TableResults, already partially evaluated by this transformation (field data and metadata)
	  * @return A new iterable of the same form, with all TableResults combined
	  */
	def assemblePartial(data : Iterable[TableResult]) : TableResult

class DefaultAssembler extends Assembler:
	def assemblePartial(data : Iterable[TableResult]) : TableResult = data.reduce(_ ++ _)

object TableTransformation:
	def fromProtobuf(transformations: Seq[table_model.TableTransformation]) : Seq[TableTransformation] = transformations.map(_.instruction match {
		case table_model.TableTransformation.Instruction.Select(expr) => SelectTransformation.fromProtobuf(expr)
		case table_model.TableTransformation.Instruction.Filter(expr) => FilterTransformation.fromProtobuf(expr)
		case _ => throw new IllegalArgumentException("Not implemented yet")
	})

/**
* Defines a transformation on a table
*/
sealed trait TableTransformation:
	val assembler : Assembler = DefaultAssembler()

	/**
	  * Checks whether this TableTransformation can be resolved with a given header
	  * 
	  * @param header A TableResultHeader instance (field metadata)
	  * @return A boolean representing whether this transformation is valid in the provided header
	  */
	def isValid(header : TableResultHeader) : Boolean 

	/**
	 * Evaluates this transformation on some provided source data
	 *
	 * @param data An iterable containing a map of field names to TableValues (field data and metadata)
	 * @return A new iterable of the same form, with the transformation applied
	 */
	def evaluate(data : TableResult) : TableResult
	
	def useDefaultAssembler : Boolean = true

	/**
	 * Returns the headers that this transformation will output
	 * 
	 * @param inputHeaders The headers being input to this transformation
	 * @return A TableResultHeader instance representing the output headers.
	 */
	def outputHeaders(inputHeaders : TableResultHeader) : TableResultHeader

	/**
	  * Returns the headers that this transformation will output when producing a partial result
	  *
	  * @param inputHeaders
	  * @return
	  */
	def outputPartialHeaders(inputHeaders : TableResultHeader) : TableResultHeader

	def protobuf : table_model.TableTransformation

object SelectTransformation:
	def fromProtobuf(select : table_model.Select) : SelectTransformation = SelectTransformation(select.fields.map(NamedFieldExpression.fromProtobuf(_))*)

final case class SelectTransformation(selectColumns : NamedFieldExpression*) extends TableTransformation:
	def isValid(header : TableResultHeader) : Boolean = Try(selectColumns.map(_.resolve(header))).isSuccess

	def evaluate(data : TableResult) : TableResult = {
		val resolved = selectColumns.map(_.resolve(data.header))
		return LazyTableResult(outputHeaders(data.header), data.rows.map(row => resolved.map(_.evaluate(row))))
	}

	def outputHeaders(inputHeaders : TableResultHeader) : TableResultHeader = TableResultHeader(selectColumns.map(_.outputTableField(inputHeaders)))

	def outputPartialHeaders(inputHeaders: TableResultHeader): TableResultHeader = outputHeaders(inputHeaders)

	lazy val protobuf = table_model.TableTransformation().withSelect(table_model.Select(selectColumns.map(_.protobuf)))

object FilterTransformation:
	def fromProtobuf(filter : table_model.Filter) : FilterTransformation = FilterTransformation(FieldComparison.fromProtobuf(filter))

final case class FilterTransformation(filter : FieldComparison) extends TableTransformation:
	def isValid(header : TableResultHeader) : Boolean = Try(filter.resolve(header)).isSuccess

	def evaluate(data: TableResult): TableResult = {
		val resolved = filter.resolve(data.header)
		return LazyTableResult(data.header, data.rows.filter(resolved.evaluate(_)))
	}

	def outputHeaders(inputHeaders : TableResultHeader) : TableResultHeader = inputHeaders

	def outputPartialHeaders(inputHeaders: TableResultHeader): TableResultHeader = outputHeaders(inputHeaders)

	lazy val protobuf = table_model.TableTransformation().withFilter(filter.protobuf)

object AggregateTransformation:
	def fromProtobuf(aggregate : table_model.Aggregate) : AggregateTransformation = AggregateTransformation(aggregate.aggregateFields.map(AggregateExpression.fromProtobuf(_))*)

final case class AggregateTransformation(aggregateColumns : AggregateExpression*) extends TableTransformation:
	class AggregateAssembler extends Assembler:
		def assemblePartial(data: Iterable[TableResult]): TableResult = {
			// Sense check: do the headers of all partial results match.
			if !data.forall(_.header == data.head.header) then throw new IllegalArgumentException("Headers of all results do not match.")
			return LazyTableResult(
				partialToFinalHeaders(data.head.header),
				// Lots of data pivoting going on here:
				// Assemble takes a header (we pick the first as we know they're all the same now)
				// We also have to give it a list of rows, but we already have a list of tables, so we append all the tables together.
				// We then flatten the results of each aggregate output to get the final table
				Seq(aggregateColumns.map(_.assemble(data.head.header)(data.map(_.rows).reduce(_ ++ _))).flatten)
			)
		}

	override val assembler = AggregateAssembler()

	def isValid(header : TableResultHeader) : Boolean = Try{aggregateColumns.map(_.resolve(header))}.isSuccess


	def evaluate(data : TableResult) : TableResult = LazyTableResult(
		// Get partial headers
		outputPartialHeaders(data.header), 
		// For each aggregate column, calculate its output, pivot all outputs to become one row, and return this
		Seq(aggregateColumns.map(_.resolve(data.header)(data.rows)).flatten)
	)

	def outputHeaders(inputHeaders : TableResultHeader) : TableResultHeader = TableResultHeader(aggregateColumns.flatMap(_.outputFinalTableFields(inputHeaders)))

	def partialToFinalHeaders(inputHeaders : TableResultHeader) : TableResultHeader = TableResultHeader(aggregateColumns.flatMap(_.outputTableFields(inputHeaders)))

	def outputPartialHeaders(inputHeaders: TableResultHeader): TableResultHeader = TableResultHeader(aggregateColumns.flatMap(_.outputPartialTableFields(inputHeaders)))

	lazy val protobuf = table_model.TableTransformation().withAggregate(table_model.Aggregate(aggregateColumns.map(_.protobuf)))

abstract final case class OrderByTransformation(orderByColumns : (FieldExpression, table_model.OrderBy.OrderByType)) extends TableTransformation

abstract final case class GroupByTransformation(groupByColumns : FieldExpression*) extends TableTransformation

abstract final case class JoinTransformation(joinType : table_model.Join.JoinType, joinTable : Table) extends TableTransformation

abstract final case class WindowTransformation(windowFunctions : Seq[FieldExpression], partitionFields : Seq[FieldExpression], orderBy : OrderByTransformation) extends TableTransformation

File: python_client/cluster_client/model/table_transformation.py
from __future__ import annotations
from typing import List, Tuple

from .field_expressions import *
from .aggregate_expressions import AggregateExpression
import cluster_client.protobuf.table_model_pb2 as protobuf_model

class TableTransformation():
    """Base class defining all transformations on a table"""
    def __init__(self):
        if type(self) == TableTransformation:
            raise NotImplementedError("TableTransformation is an abstract class and cannot be instantiated directly.")
    
    def to_protobuf() -> protobuf_model.TableTransformation: 
        raise NotImplementedError("TableTransformation abstract class cannot be converted to protobuf.")

class SelectTransformation(TableTransformation):
    """Selects any number of columns from the tables, and applies any given field expressions"""
    def __init__(self, *select_columns : NamedFieldExpression):
        self.select_columns = select_columns

    def to_protobuf(self) -> protobuf_model.TableTransformation:
        return protobuf_model.TableTransformation(select=protobuf_model.Select(fields=[col.to_named_protobuf() for col in self.select_columns]))

    def __str__(self):
        col_data = ""
        for col in self.select_columns:
            col_data += str(col) + ", "
        col_data = col_data[:-2]
        return f"Select({col_data})"

class FilterTransformation(TableTransformation):
    """Filters the table according to some Filter criteria"""
    def __init__(self, filter: FieldComparison):
        self.filter = filter
    
    def to_protobuf(self) -> protobuf_model.TableTransformation:
        return protobuf_model.TableTransformation(filter=self.filter.to_protobuf())

    def __str__(self):
        filter_data = ""

        for filter in self.filter:
            filter_data += str(filter) + ", "
        filter_data = filter_data[:-2]
        return f"Filter({filter_data})"

class JoinTransformation(TableTransformation):
    def __init__(self, join_type : str,  join_table_name : str):
        self.join_table_name = join_table_name

        try:
            self.join_type = protobuf_model.Join.JoinType.Value(join_type.upper())
        except ValueError:
            raise ValueError(f"Invalid join type provided {join_type}")

    def to_protobuf(self) -> protobuf_model.TableTransformation:
        return protobuf_model.TableTransformation(join=
            protobuf_model.Join(join_type=self.join_type, table_name=self.join_table_name)
        )

    def __str__(self):
        return f"Join({self.join_type}, {self.join_table_name})"

class GroupByTransformation(TableTransformation):
    """Groups future calculations by the provided expressions"""
    def __init__(self, *group_by_columns : FieldExpression):
        self.group_by_columns = group_by_columns

    def to_protobuf(self) -> protobuf_model.TableTransformation:
        return protobuf_model.TableTransformation(group_by=protobuf_model.GroupBy(fields=[col.to_protobuf() for col in self.group_by_columns]))

    def __str__(self):
        group_data = ""
        for group_by_column in self.group_by_columns:
            group_data += str(group_by_column) + ", "
        group_data = group_data[:-2]
        return f"Filter({group_data})"

class AggregateTransformation(TableTransformation):
    """Aggregates the data in the table by the provided expressions"""
    def __init__(self, *aggregate_columns : AggregateExpression):
        self.aggregate_columns = aggregate_columns

    def to_protobuf(self) -> protobuf_model.TableTransformation:
        return protobuf_model.TableTransformation(aggregate=protobuf_model.Aggregate(aggregate_fields=[col.to_protobuf() for col in self.aggregate_columns]))

    def __str__(self):
        aggregate_data = ""
        for aggregate_column in self.aggregate_columns:
            aggregate_data += str(aggregate_column) + ", "
        aggregate_data = aggregate_data[:-2]
        return f"Aggregate({aggregate_data})"

class OrderByTransformation(TableTransformation):
    """Orders the output results by the provided expressions"""
    def __init__(self, *order_by_columns : Tuple(FieldExpression, str)):
        converted_pairs = []
        for expression, order_by_type in order_by_columns:
            try:
                converted_pairs.append((expression, protobuf_model.OrderBy.OrderByType.Value(order_by_type)))
            except ValueError:
                raise ValueError(f"Invalid order by type provided {order_by_type}")

        self.order_by_columns = converted_pairs

    def to_protobuf(self) -> protobuf_model.TableTransformation:
        return protobuf_model.TableTransformation(order_by=protobuf_model.Order(order_by_fields=[
            protobuf_model.Order.OrderByField(field=expression.to_protobuf(), order_by_type=order_by_type) for expression, order_by_type in self.order_by_columns
        ]))

    def __str__(self):
        order_by_data = ""
        for order_by_column in self.order_by_columns:
            order_by_data += str(order_by_column) + ", "
        order_by_data = order_by_data[:-2]
        return f"OrderBy({order_by_data})"

class WindowTransformation(TableTransformation):
    """Performs a calculation over a window function"""
    def __init__(self, window_functions : List[FieldExpression], partition_fields : List[str], order_by : List(Tuple(FieldExpression, str)) = None):
        self.window_functions = window_functions
        self.partition_fields = partition_fields
        if order_by is not None:
            self.order_by = OrderByTransformation(order_by)

    def to_protobuf(self) -> protobuf_model.TableTransformation:
        return protobuf_model.TableTransformation(window=
            protobuf_model.Window(
                window_function=[window_function.to_protobuf() for window_function in self.window_functions],
                partition_fields=self.partition_fields,
                order_by_fields=self.order_by
            )
        )

    def __str__(self):
        window_data = ""
        for window_function in self.window_functions:
            window_data += str(window_function) + ", "
        window_data = window_data[:-2]

        partition_data = ""
        for partition_fields in self.partition_fields:
            partition_data += str(partition_fields) + ", "
        partition_data = partition_data[:-2]

        order_by_data = ""
        if self.order_by is not None:
            order_by_data = str(self.order_by)

        return f"Window({window_data}) over ({partition_data}) {order_by_data}"
